bootstrap.go

```diff
@@ -147,38 +154,107 @@ func checkBootstrapped(s Session) (bool, error) {
-			return false, nil
-		}
-		return false, errors.Trace(err)
+		return types.Datum{}, errors.Trace(err)
 	}
-
 	if len(rs) != 1 {
-		return false, errors.New("Wrong number of Recordset")
+		return types.Datum{}, errors.New("Wrong number of Recordset")
 	}
 	r := rs[0]
 	row, err := r.Next()
 	if err != nil || row == nil {
-		return false, errors.Trace(err)
+		return types.Datum{}, errors.Trace(err)
 	}
+	return row.Data[0], nil
+}
 
-	isBootstrapped := row.Data[0].GetString() == bootstrappedVarTrue
+func checkBootstrappedVar(s Session) (bool, error) {
+	d, err := getTiDBVar(s, bootstrappedVar)
+	if err != nil {
+		if infoschema.ErrTableNotExists.Equal(err) {
+			return false, nil
+		}
+		return false, errors.Trace(err)
+	}
+	isBootstrapped := d.GetString() == bootstrappedVarTrue
 	if isBootstrapped {
 		// Make sure that doesn't affect the following operations.
-
 		if err = s.CommitTxn(); err != nil {
 			return false, errors.Trace(err)
 		}
 	}
-
 	return isBootstrapped, nil
 }
 
+// When the system is boostrapped by low version TiDB server, we should do some upgrade works.
+// For example, add new system variables into mysql.global_variables table.
+func upgrade(s Session) error {
+	ver, err := getBootstrapVersion(s)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if ver >= currentBootstrapVersion {
+		// It is already bootstrapped/upgraded by a higher version TiDB server.
+		if err1 := s.CommitTxn(); err1 != nil {
+			// Make sure that doesn't affect the following operations.
+			return errors.Trace(err1)
+		}
+		return nil
+	}
+	// Do upgrade works.
+	var sql string
+	if ver < 2 {
+		// Version 2 add two system variable for DistSQL concurrency controling.
+		// Insert distsql related system variable.
+		distSQLVars := []string{variable.DistSQLScanConcurrencyVar, variable.DistSQLJoinConcurrencyVar}
+		values := make([]string, 0, len(distSQLVars))
+		for _, v := range distSQLVars {
+			value := fmt.Sprintf(`("%s", "%s")`, v, variable.SysVars[v].Value)
+			values = append(values, value)
+		}
+		sql := fmt.Sprintf("INSERT IGNORE INTO %s.%s VALUES %s;", mysql.SystemDB, mysql.GlobalVariablesTable,
+			strings.Join(values, ", "))
+		mustExecute(s, sql)
+	}
+	// Update bootstrap version.
+	sql = fmt.Sprintf(`INSERT INTO %s.%s VALUES ("%s", "%d", "TiDB bootstrap version.") ON DUPLICATE KEY UPDATE VARIABLE_VALUE="%d"`,
+		mysql.SystemDB, mysql.TiDBTable, tidbServerVersionVar, currentBootstrapVersion, currentBootstrapVersion)
+	mustExecute(s, sql)
+	_, err = s.Execute("COMMIT")
+	if err != nil {
+		time.Sleep(1 * time.Second)
+		// Check if TiDB is already upgraded.
+		v, err1 := getBootstrapVersion(s)
+		if err1 != nil {
+			log.Fatal(err1)
+		}
+		if v >= currentBootstrapVersion {
+			return nil
+		}
+		log.Errorf("[Upgrade] upgrade from %d to %d error", ver, currentBootstrapVersion)
+		log.Fatal(err)
+	}
+	return nil
+}
+
+// Get bootstrap version from mysql.tidb table;
+func getBootstrapVersion(s Session) (int64, error) {
```

Comment:
Can we use this function to replace `checkBootstrappedVar`?
version 0 means not bootstrapped.